Add OPT

younesbelkada opened the pull request:
# What does this PR do?

A PR to add OPT-350-m model on transformers library :hugs: 
We tested the logits and the generation script and the logits on the DGX and we match the results from [metaseq](https://github.com/patrickvonplaten/metaseq/blob/main/README.md)

## TO DOs

- [ ] code lint + integration tests
- [ ] discuss how to import the logits for the hardcoded tests
- [ ] correct documentation


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc  @LysandreJik  @patrickvonplaten @stas00 @sgugger

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

younesbelkada committed:
First version - OPT model

younesbelkada committed:
Final changes

- putting use cache to False

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1821 @@
+                v_proj_weight=self.v_proj.weight,
+            )
+
+        # bsz, tgt_len, _ = hidden_states.size()
```
@younesbelkada does it not work with the commented out code or is it just "cleaner" to use PyTorch's `multi_head_attention_forward`?

younesbelkada replied:
It is what they used in their implementation and worked to obtain the same logits (also I found it a bit cleaner)

younesbelkada replied:
resolved in 0484ca10e74a245cd878166e3384eb51a0d97042

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
````diff
@@ -0,0 +1,1821 @@
+    'PG&E scheduled the blackouts in response to forecasts for high winds amid dry conditions'
+    ```
+
+    Mask filling example:
````
Let's delete this docstring

younesbelkada replied:
The whole `OPT_INPUTS_DOCSTRING` ?

patrickvonplaten replied:
Sorry I meant just the `Mask filling example` :-)

younesbelkada replied:
Oh okay I see! Resolved in 5e8e2f599be59d9a3f9692c93318b6a091005a46

patrickvonplaten commented on src/transformers/models/opt/tokenization_opt.py:
```diff
@@ -0,0 +1,417 @@
+# coding=utf-8
```
Let's delete this file

younesbelkada replied:
resolved in b931db8efbe7e79c7da2114017fe587265e2007e

patrickvonplaten commented on src/transformers/models/opt/tokenization_opt_fast.py:
```diff
@@ -0,0 +1,303 @@
+# coding=utf-8
```
Let's delete this file

The tokenizer is a GPT2-like tokenizer

patrickvonplaten replied:
We can reuse its code

younesbelkada replied:
resolved in b931db8efbe7e79c7da2114017fe587265e2007e

patrickvonplaten commented on tests/models/opt/test_tokenization_opt.py:
```diff
@@ -0,0 +1,185 @@
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
Let's delete this

younesbelkada replied:
resolved in 1e21983a4ed08ca2c91d2d218e5048ba50998efa

patrickvonplaten commented on tests/models/opt/test_embeddings_opt.py:
```diff
@@ -0,0 +1,66 @@
+        pipeline,
+    )
+
+class OPTEmbeddingsTest(unittest.TestCase):
```
Can we add this as an integration test under `test_modeling_opt.py`?

patrickvonplaten replied:
+ we should also add a generation test

younesbelkada replied:
resolved here 842727999cb6d7a298963021dcd5daafb98d2453

younesbelkada committed:
few changes

- remove commented block

younesbelkada committed:
few changes

- remove unecessary files

younesbelkada committed:
fix style issues

younesbelkada committed:
few changes

- remove a test file
- added the logits test

younesbelkada committed:
Update src/transformers/models/auto/tokenization_auto.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

younesbelkada commented:
Also I have noticed that here: https://github.com/younesbelkada/transformers/blob/13632212d5bacc83bb9a9b8b6816a70bde865dbc/src/transformers/models/opt/modeling_opt.py#L249 
Initially in our implementation the attention mask has a shape `batch_size, 1, seq_len, seq_len` whereas in the metaseq implementation the masks is identitcal across the whole batch therefore has a shape of `seq_len, seq_len`. Is it a common practice that the attention mask is identical across the whole batch ?

younesbelkada committed:
add gen tests

patrickvonplaten commented:
> identitcal

I shouldn't but I think for causal LM it's fine to do this since the attention_mask is usually just the causal mask which is the same for every batch. Later tokens can never attend to earlier tokens when there is a causal mask, so it's often good enough to use that even when tokens are padded

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1708 @@
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        key_value_states: Optional[torch.Tensor] = None,
```
guess we can delete this arg

younesbelkada replied:
actually we could delete more args, did several changes in this commit: be0e434a55015638d5de9821748d4cbf08e54d80

younesbelkada committed:
few changes

- rm mask filling example on docstring

younesbelkada committed:
few changes

- remove useless args

younesbelkada commented:
Thank you very much for the explanation regarding the attention mask
Since I took the template from BART I think that it considered this attention strategy by default

younesbelkada commented:
I am facing [this CI issue]() when running `make repo-consistency` and I can't figure out what is happening.. Do you have any idea by change what could cause this issue?
![Screenshot from 2022-05-05 00-00-12](https://user-images.githubusercontent.com/49240599/166832823-e85fc11b-9004-49ef-a15d-e18a010368ff.png)

younesbelkada commented:
Linking this issuehttps://github.com/facebookresearch/metaseq/issues/31 
to understand how to load properly sharded models

suchenzang commented:
Crashing the party here (apologies for not having more context on this codebase): is it normal for model imports to require >8k LOC added? Or is there something in particular about OPT models that requires all this additional glue code? 

If this is is OPT's fault - is there anything we can do (on the [metaseq](https://github.com/facebookresearch/metaseq) side) to bridge this gap?

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+        assert torch.equal(attention_mask[0, 0, :, :], attention_mask[-1, 0, :, :])
+        attention_mask = attention_mask[0, 0, :, :]
+
+        return F.multi_head_attention_forward(
```
We need to add a:

```
average_attn_weights=False
```

here **or (which might be preferred)** write out the attention function ourselves like we do for other models. E.g., like we do for BartDecoder.

@younesbelkada doesn't this attention functionality: https://github.com/huggingface/transformers/blob/99289c08a1b16a805dd4ee46de029e9fd23cba3d/src/transformers/models/bart/modeling_bart.py#L188 work? Why delete it here?

patrickvonplaten replied:
cc @LysandreJik @sgugger what do you think?

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+
+        # create key padding mask
+        # HACK : to fix
+        assert torch.equal(attention_mask[0, 0, :, :], attention_mask[-1, 0, :, :])
```
We need to clean this up

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+    OPT_START_DOCSTRING,
+)
+# Copied from transformers.models.bart.modeling_bart.BartForQuestionAnswering with BART->OPT,Bart->OPT
+class OPTForQuestionAnswering(OPTPretrainedModel):
```
Let's delete this whole class

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+    OPT_START_DOCSTRING,
+)
+# Copied from transformers.models.bart.modeling_bart.BartForSequenceClassification with BART->OPT,Bart->OPT
+class OPTForSequenceClassification(OPTPretrainedModel):
```
Let's delete this whole class

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+    "The OPT Model with a language modeling head. Can be used for summarization.", OPT_START_DOCSTRING
+)
+# Copied from transformers.models.bart.modeling_bart.BartForConditionalGeneration with BART->OPT,Bart->OPT
+class OPTForConditionalGeneration(OPTPretrainedModel):
```
doesn't exist either - OPT is a decoder-only model. Let's delete it

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+
+
+# Copied from transformers.models.bart.modeling_bart.BartForCausalLM with Bart->OPT,facebook/bart-base->
+class OPTForCausalLM(OPTPretrainedModel):
```
Nice - that's the important class we need

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+        if output_attentions:
+            outputs += (self_attn_weights, cross_attn_weights)
+
+        # if use_cache:
```
? why comment out here?

patrickvonplaten replied:
We need the cache to work no?

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+                returned tensors for more detail.
+        """
+
+        # hidden_states = hidden_states.permute(1, 0, 2)
```
?

patrickvonplaten replied:
why comment out?

younesbelkada replied:
Yeah it should be deleted sorry for the confusion

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+        return dummy_inputs
+
+
+# Copied from transformers.models.bart.modeling_bart.PretrainedBartModel with Bart->OPT
```
Let's delete this class as well, it's deprecated

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+"""
+
+OPT_GENERATION_EXAMPLE = r"""
+    Summarization example:
```
OPT can't do summarization, let's give it a text generation example

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+            - 0 for tokens that are **masked**.
+
+            [What are attention masks?](../glossary#attention-mask)
+        decoder_input_ids (`torch.LongTensor` of shape `(batch_size, target_sequence_length)`, *optional*):
```
doesn't exist

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+            For translation and summarization training, `decoder_input_ids` should be provided. If no
+            `decoder_input_ids` is provided, the model will create this tensor by shifting the `input_ids` to the right
+            for denoising pre-training following the paper.
+        decoder_attention_mask (`torch.LongTensor` of shape `(batch_size, target_sequence_length)`, *optional*):
```
doesn't exist

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+            - 1 indicates the head is **not masked**,
+            - 0 indicates the head is **masked**.
+
+        decoder_head_mask (`torch.Tensor` of shape `(decoder_layers, decoder_attention_heads)`, *optional*):
```
to delete

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+            - 1 indicates the head is **not masked**,
+            - 0 indicates the head is **masked**.
+
+        cross_attn_head_mask (`torch.Tensor` of shape `(decoder_layers, decoder_attention_heads)`, *optional*):
```
doesn't exist

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+            `(batch_size, sequence_length, hidden_size)`, *optional*): Optionally, instead of passing `input_ids` you
+            can choose to directly pass an embedded representation. This is useful if you want more control over how to
+            convert `input_ids` indices into associated vectors than the model's internal embedding lookup matrix.
+        decoder_inputs_embeds (`torch.FloatTensor` of shape `(batch_size, target_sequence_length, hidden_size)`, *optional*):
```
doesn't exist

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+    return (torch.cumsum(mask, dim=1).type_as(mask) * mask).long() + padding_idx
+
+
+class LearnedPositionalEmbedding(nn.Embedding):
```
This needs a prefix

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+            self.padding_idx is None
+        ), "If positions is pre-computed then padding_idx should not be set."
+
+        # we cannot use incremental state here because we must be aware of
```
I don't understand this comment

patrickvonplaten replied:
Are you referring to the case when you use `past_key_values` states?

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1681 @@
+
+
+# Copied from transformers.models.bart.modeling_bart.BartLearnedPositionalEmbedding with Bart->OPT
+# class OPTLearnedPositionalEmbedding(nn.Embedding):
```
? Do we keep this or not?

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,397 @@
+            Number of decoder layers.
+        encoder_attention_heads (`int`, *optional*, defaults to 16):
+            Number of attention heads for each attention layer in the Transformer encoder.
+        decoder_attention_heads (`int`, *optional*, defaults to 16):
```
doesn't exist

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,397 @@
+            Number of attention heads for each attention layer in the Transformer encoder.
+        decoder_attention_heads (`int`, *optional*, defaults to 16):
+            Number of attention heads for each attention layer in the Transformer decoder.
+        decoder_ffn_dim (`int`, *optional*, defaults to 4096):
```
doesn't exist

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,397 @@
+        self,
+        vocab_size=50272,  # TODO check the real value
+        max_position_embeddings=2048,
+        decoder_layers=24,
```
doesn't exist

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,397 @@
+        vocab_size=50272,  # TODO check the real value
+        max_position_embeddings=2048,
+        decoder_layers=24,
+        decoder_ffn_dim=4096,
```
doesn't exist

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,397 @@
+        max_position_embeddings=2048,
+        decoder_layers=24,
+        decoder_ffn_dim=4096,
+        decoder_attention_heads=16,
```
doesn't exist

patrickvonplaten commented:
> Crashing the party here (apologies for not having more context on this codebase): is it normal for model imports to require >8k LOC added? Or is there something in particular about OPT models that requires all this additional glue code?
> 
> If this is is OPT's fault - is there anything we can do (on the [metaseq](https://github.com/facebookresearch/metaseq) side) to bridge this gap?

Hey @suchenzang,

Cool that you're taking a look here! There is lot of code that was copied in a first pass from Bart that should be removed. This will make the code much more concise. Also note that we'll add the model both in TF and Flax as well. 

On the other hand, you guys are the authors of OPT, so you should lead the addition of OPT to Transformers if you want to. Would you like to open a PR yourself instead that adds the model instead of us doing it? We took the initiative here because of timing, but if you would like to take over the integration, please let us know :-)

younesbelkada committed:
some changes

- more tests should pass now
- needs to clean more
- documentation still needs to be done

younesbelkada committed:
fix code quality

suchenzang commented:
> On the other hand, you guys are the authors of OPT, so you should lead the addition of OPT to Transformers if you want to. Would you like to open a PR yourself instead that adds the model instead of us doing it? We took the initiative here because of timing, but if you would like to take over the integration, please let us know :-)

@patrickvonplaten Not at all - love seeing this initiative (and we likely don't have bandwidth to take on integration work)!  We have things to clean up on the metaseq side too, so I'm generally looking for feedback on this front as more folks try out our models.

mrseeker commented:
Just out of curiosity, and to prevent duplicate work: Since fairseq models can be converted 1 on 1 to the XGLM model type, would that be possible with this one too or is this model too different to warrant a totally new model type?
Secondly, this is just the 350m model. Are other variants of this model going to be supported too? Hate to see this merged without full OPT model support.

younesbelkada committed:
major changes

- change attention architecture to BART-like
- modify some tests
- style fix

ArthurZucker commented on docs/source/en/model_doc/opt.mdx:
```diff
@@ -0,0 +1,128 @@
+[[autodoc]] OPTConfig
+    - all
+
+## OPTTokenizer
```
Should the OPTTokenizer be removed to pass the tests? On error is on serialising the .mdx file

younesbelkada committed:
rm useless classes

- remove opt for:
- QA
- cond generation
- seq classif

ArthurZucker commented on src/transformers/__init__.py:
```diff
@@ -2646,6 +2672,7 @@
     from .models.mt5 import MT5Config
     from .models.nystromformer import NYSTROMFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP, NystromformerConfig
     from .models.openai import OPENAI_GPT_PRETRAINED_CONFIG_ARCHIVE_MAP, OpenAIGPTConfig, OpenAIGPTTokenizer
+    from .models.opt import OPTConfig, OPTTokenizer
```
Tokenizer not implemented, might cause 
```
tests/utils/test_file_utils.py:26: in <module>
 from transformers import *
```

ArthurZucker commented on src/transformers/__init__.py:
```diff
@@ -238,6 +238,7 @@
         "NystromformerConfig",
     ],
     "models.openai": ["OPENAI_GPT_PRETRAINED_CONFIG_ARCHIVE_MAP", "OpenAIGPTConfig", "OpenAIGPTTokenizer"],
+    "models.opt": ["OPTConfig", "OPTTokenizer"],
```
No tokenizer file, should "OPTTokenizer be included?

ArthurZucker reviewed (commented):
Hey, as we talked today, here is what I found, might help 👍🏻

ArthurZucker committed:
Removed autodoc calls to non-existant classes

TOkenizers are not implemented

ArthurZucker commented on src/transformers/__init__.py:
```diff
@@ -483,6 +484,7 @@
     _import_structure["models.mpnet"].append("MPNetTokenizerFast")
     _import_structure["models.mt5"].append("MT5TokenizerFast")
     _import_structure["models.openai"].append("OpenAIGPTTokenizerFast")
+    _import_structure["models.opt"].append("OPTTokenizerFast")
```
```suggestion
 
```

ArthurZucker commented on src/transformers/__init__.py:
```diff
@@ -2870,6 +2894,7 @@
         from .models.mpnet import MPNetTokenizerFast
         from .models.mt5 import MT5TokenizerFast
         from .models.openai import OpenAIGPTTokenizerFast
+        from .models.opt import OPTTokenizerFast
```
```suggestion
  
```

ArthurZucker commented on src/transformers/models/auto/tokenization_auto.py:
```diff
@@ -107,6 +107,7 @@
             ("blenderbot", ("BlenderbotTokenizer", "BlenderbotTokenizerFast")),
             ("tapex", ("TapexTokenizer", None)),
             ("bart", ("BartTokenizer", "BartTokenizerFast")),
+            ("opt", ("GPT2Tokenizer", "GPT2TokenizerFast")),
```
```suggestion
```

ArthurZucker commented on src/transformers/models/auto/modeling_tf_auto.py:
```diff
@@ -118,6 +120,7 @@
         ("albert", "TFAlbertForMaskedLM"),
         ("marian", "TFMarianMTModel"),
         ("bart", "TFBartForConditionalGeneration"),
+        ("opt", "TFOPTForConditionalGeneration"),
```
```suggestion
```

younesbelkada committed:
Update src/transformers/__init__.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
Update src/transformers/__init__.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
Update src/transformers/models/auto/modeling_tf_auto.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

ArthurZucker committed:
Replaced OPTTokeniser with GPT2 tokenizer

ArthurZucker committed:
added GPT2Tokenizer.from_pretrained("patrickvonplaten/opt_gpt2_tokenizer")

ArthurZucker committed:
Removed OPTTokenizer

ArthurZucker committed:
make style

ArthurZucker committed:
Make style replaces

``` ...).unsqueeze(```
by
``` >>>).unsqueeze(```

ArthurZucker committed:
make repo consistency

ArthurZucker committed:
Removed PretrainedOPTModel

ArthurZucker committed:
fix opt.mdx removed other heads

ArthurZucker committed:
fix init, removed 3 heads

ArthurZucker committed:
removed heads

ArthurZucker committed:
finished cleaning head

ArthurZucker committed:
removed seauence classif and question answering

ArthurZucker committed:
removed unused imports

ArthurZucker committed:
removed useless dummy object for QA, SC and CG

ArthurZucker committed:
removed tests for removed useless dummy object for QA, SC and CG

ArthurZucker committed:
Removed head_mask using encoder layers which don't exist

ArthurZucker committed:
fixed test

ArthurZucker committed:
fix line

ArthurZucker committed:
added OPT to toctree

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17088). All of your documentation changes will be reflected on that endpoint.

ArthurZucker committed:
Updated model path with pushed weigths

ArthurZucker committed:
fix model path

ArthurZucker committed:
Merge branch 'main' of https://github.com/huggingface/transformers into opt-350-m

ArthurZucker committed:
fixed code quality

ArthurZucker committed:
fixed embeddings and generation tests

ArthurZucker committed:
update paths

ArthurZucker committed:
clean comments

ArthurZucker committed:
removed OPTClassificationHead for sentence classification

ArthurZucker committed:
renamed hidden layer

ArthurZucker committed:
renamed num layers to standard num_hidden_layers

ArthurZucker committed:
num_attention_heads fix

younesbelkada committed:
changes for 125m

younesbelkada committed:
Merge branch 'opt-350-m' of https://github.com/younesbelkada/transformers into opt-350-m

younesbelkada committed:
add first version for 125m

younesbelkada committed:
add first version - flax

younesbelkada committed:
Merge branch 'opt-350-m' of https://github.com/younesbelkada/transformers into opt-350-m

younesbelkada committed:
add new version

younesbelkada committed:
Merge branch 'opt-350-m' of https://github.com/younesbelkada/transformers into opt-350-m

ArthurZucker committed:
causal LM output

ArthurZucker committed:
Merge branch 'opt-350-m' of github.com:younesbelkada/transformers into opt-350-m

ArthurZucker committed:
replace output type with BaseModelOutputWithPastAndCrossAttentions

ArthurZucker committed:
revert working config from 150m to 350m

ArthurZucker committed:
clean

ArthurZucker committed:
removed decoder input ids

ArthurZucker committed:
fixed embed dim

ArthurZucker committed:
more embed_dim issues

ArthurZucker committed:
make style + removed enc_dec test

ArthurZucker committed:
update falx model

ArthurZucker committed:
removed troublesome copy

ArthurZucker committed:
added is_encoder_decoder=False to config

ArthurZucker committed:
added set_input emb fuinction to model class

ArthurZucker committed:
requires torch on embed test

ArthurZucker committed:
use head mask instead of decoder head mask input param solves a test

ArthurZucker committed:
8 test remaining, update

ArthurZucker committed:
Updated create_and_check_decoder_model_past_large_inputs

ArthurZucker committed:
Make style

ArthurZucker committed:
update op tokenizer with condition

ArthurZucker committed:
make style

patrickvonplaten committed:
See if I can push

patrickvonplaten committed:
some clean up

patrickvonplaten committed:
remove linear head hack

patrickvonplaten committed:
save intermediate

patrickvonplaten committed:
save correct attention

patrickvonplaten committed:
add copied from from bart

patrickvonplaten committed:
Merge branch 'main' of https://github.com/huggingface/transformers into opt-350-m

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,997 @@
+        )
+        self.dropout = config.dropout
+        self.activation_fn = ACT2FN["relu"]
+        # self.activation_fn = ACT2FN[config.activation_function]
```
```suggestion
        self.activation_fn = ACT2FN[config.activation_function]
```

Let's change `config.activation_function` to `"relu"` in the config

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
````diff
@@ -0,0 +1,997 @@
+    Generation example:
+
+    ```python
+    >>> from transformers import GPT2Tokenizer, OPTForCausalLM
````
Let's make this a nice example with the `350m` checkpoint.

sgugger replied:
```suggestion
    >>> from transformers import AutoTokenizer, AutoModelForCausalLM
```
Using the auto APIs is better, so that the user doesn't have to remember it's the GPT2 tokenizer for this model

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,997 @@
+    >>> TEXTS_TO_GENERATE = "Hey, are you consciours? Can you talk to me?" "Hi there, my name is Barack"
+    >>> inputs = tokenizer([TEXTS_TO_GENERATE], max_length=1024, return_tensors="pt")
+
+    >>> # Generate Summary
```
```suggestion
    >>> # Generate
```

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,997 @@
+    >>> inputs = tokenizer([TEXTS_TO_GENERATE], max_length=1024, return_tensors="pt")
+
+    >>> # Generate Summary
+    >>> summary_ids = model.generate(inputs["input_ids"], num_beams=2, min_length=0, max_length=20)
```
```suggestion
    >>> generate_ids = model.generate(inputs["input_ids"], num_beams=2, min_length=0, max_length=20)
```

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,997 @@
+            Indices can be obtained using [`OPTTokenizer`]. See [`PreTrainedTokenizer.encode`] and
+            [`PreTrainedTokenizer.__call__`] for details.
+
+            OPT uses the `eos_token_id` as the starting token for `decoder_input_ids` generation. If `past_key_values`
```
```suggestion
```

And all following lines

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,997 @@
+        self.max_target_positions = config.max_position_embeddings
+        self.share_input_output_embed = config.share_input_output_embed
+        self.embed_scale = math.sqrt(config.d_model) if config.scale_embedding else 1.0
+        self.cross_self_attention = False  # TODO add it on the config
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,997 @@
+        self.embed_tokens = nn.Embedding(config.vocab_size, config.embed_dim, self.padding_idx)
+
+        if self.padding_idx is not None:
+            num_embeddings = config.max_position_embeddings + self.padding_idx + 1
```
```suggestion
            num_embeddings = config.max_position_embeddings + 2
```
Let's force it to be 2

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,997 @@
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        # If the user passed a tuple for encoder_outputs, we wrap it in a BaseModelOutputWithPastAndCrossAttentions when return_dict=True
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,997 @@
+        >>> from transformers import OPTTokenizer, OPTForCausalLM
+
+        # this needs fixing
+        >>> tokenizer = OPTTokenizer.from_pretrained("")
```
Let's make a better example here

patrickvonplaten commented on tests/models/opt/test_modeling_opt.py:
```diff
@@ -0,0 +1,494 @@
+
+
+@require_torch
+class OPTHeadTests(unittest.TestCase):
```
Let's delete the whole class here

patrickvonplaten commented on tests/models/opt/test_modeling_opt.py:
```diff
@@ -0,0 +1,494 @@
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_decoder_model_past_large_inputs(*config_and_inputs)
+
+    # OPTForSequenceClassification does not support inputs_embeds
```
```suggestion
```

patrickvonplaten commented on tests/models/opt/test_modeling_opt.py:
```diff
@@ -0,0 +1,494 @@
+        self.path_model = "ArthurZ/opt-350m"
+        # self.path_logits_meta = "/home/younes/Desktop/Work/metaseq-conversion/logits_metaseq_gpt2_tokenizer.p"  # TODO add the logits somewhere?
+
+    def test_generation(self):
```
Let's add 3 more generation tests here with the 125m, 350m and 1300m checkpoint that all show sensible outputs for generation

ArthurZucker replied:
Good idea, working on it

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,381 @@
+        )
+
+        # ensure backward compatibility for OPT CNN models
+        if self.forced_bos_token_id is None and kwargs.get("force_bos_token_to_be_generated", False):
```
Let's remove this

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,381 @@
+            )
+
+
+class OPTOnnxConfig(OnnxSeq2SeqConfigWithPast):
```
Delete this and all of the following code

patrickvonplaten reviewed (commented):
Cool progress @ArthurZucker , let's try to finish the PyTorch modeling file tomorrow :-)

patrickvonplaten commented:
Here the checkpoints that should be converted: https://huggingface.co/patrickvonplaten/opt_metaseq_6700m/tree/main/model

ArthurZucker committed:
Update src/transformers/models/opt/modeling_opt.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ArthurZucker committed:
fix part of the reviewss
Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ArthurZucker committed:
Merge pull request #2 from younesbelkada/opt_branch/opt-350-m

fix part of the reviews

patrickvonplaten committed:
same changes in naming / conversion

patrickvonplaten committed:
correct mask

patrickvonplaten committed:
more fixes

ArthurZucker committed:
delete FlaxOPT and TfOPT

ArthurZucker committed:
clean traces of Flax and Tf

patrickvonplaten committed:
fix mask

patrickvonplaten committed:
Merge branch 'opt-350-m' of https://github.com/younesbelkada/transformers into opt-350-m

ArthurZucker committed:
fixed positionnal embedding length when past key value is provoded

patrickvonplaten committed:
get 125m, 6.7b to work

patrickvonplaten committed:
Merge branch 'opt-350-m' of https://github.com/younesbelkada/transformers into opt-350-m

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1004 @@
+        residual = hidden_states
+
+        # Self Attention
+        # TODO(ArthurZ) could you add a flag to the config "do_layer_norm_before = True/False" to
```
@ArthurZucker here we probably need a flag in the config to differentiate between the 350m and the other architectures

ArthurZucker replied:
Okay

ArthurZucker committed:
Added do_layer_norm

ArthurZucker committed:
solved mismatch in load dictionnary

ArthurZucker committed:
clean up preapre opt input dict

ArthurZucker committed:
fixed past key value as bool

ArthurZucker committed:
fix previus

ArthurZucker committed:
fixed return dict False tuple issue

ArthurZucker committed:
All tests are passing

ArthurZucker committed:
Make style

ArthurZucker committed:
Ignore OPTDecoder non tested

ArthurZucker committed:
make fix-copies

ArthurZucker committed:
make repo consistency

ArthurZucker committed:
small fix

ArthurZucker committed:
removed uselss @torch.no_grad decorator

ArthurZucker committed:
make styl;e

ArthurZucker committed:
fix previous opt test

ArthurZucker committed:
style

ArthurZucker committed:
make style

ArthurZucker committed:
added opt documentation

ArthurZucker committed:
update OPT_PRETRAINED_MODEL_ARCHIVE_LIST

patrickvonplaten commented on docs/source/en/model_doc/opt.mdx:
```diff
@@ -0,0 +1,54 @@
+
+If you want to use the `opt-350m`, you need to set the `do_layer_norm_before` parameter to `False`. 
+
+This model was contributed by [Patrick Von Platen](https://huggingface.co/patrickvonplaten), [Arthur Zucker](https://huggingface.co/ArthurZ) and [Younes Belkada](https://huggingface.co/ybelkada),.
```
```suggestion
This model was contributed by [Arthur Zucker](https://huggingface.co/ArthurZ) and [Younes Belkada](https://huggingface.co/ybelkada), [Patrick Von Platen](https://huggingface.co/patrickvonplaten).
```

You guys wrote the code ;-)

patrickvonplaten commented on docs/source/en/model_doc/opt.mdx:
```diff
@@ -0,0 +1,54 @@
+
+Tips:
+
+If you want to use the `opt-350m`, you need to set the `do_layer_norm_before` parameter to `False`. 
```
```suggestion
```

This is done automatically in the config. The comment would confuse users IMO

patrickvonplaten commented on docs/source/en/model_doc/opt.mdx:
```diff
@@ -0,0 +1,54 @@
+## Overview
+
+The OPT model was proposed in [Open Pre-trained Transformer Language Models](https://arxiv.org/pdf/2205.01068) by Meta AI.
+OPT is a serie of pretrained decoder only LM which perform similarly to GPT3. 
```
```suggestion
OPT is a series of open-sourced large causal language models which perform similar in performance to GPT3. 
```

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,140 @@
+logger = logging.get_logger(__name__)
+
+OPT_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "": "https://huggingface.co//resolve/main/config.json",
```
Could you use one of those links: https://huggingface.co/facebook/opt-125m ?

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,140 @@
+        decoder_start_token_id=2,
+        forced_eos_token_id=2,
+        output_projection=True,
+        do_layer_norm_before=True,
```
Docstring for this?

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,140 @@
+        self.layerdrop = layerdrop
+        self.use_cache = use_cache
+        self.scale_embedding = scale_embedding  # scale factor will be sqrt(d_model) if True
+        self.output_projection = output_projection
```
Don't think that this is used

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,140 @@
+        eos_token_id=2,
+        decoder_start_token_id=2,
+        forced_eos_token_id=2,
+        output_projection=True,
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,140 @@
+        activation_dropout=0.0,
+        init_std=0.02,
+        scale_embedding=False,
+        share_input_output_embed=True,
```
not used

patrickvonplaten replied:
not we have a general param for this in `PretrainedConfig` which is called `tie_word_embeddings` or something like that

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,140 @@
+        self.num_attention_heads = num_attention_heads
+        self.word_embed_proj_dim = word_embed_proj_dim
+        self.ffn_dim = ffn_dim
+        self.share_input_output_embed = share_input_output_embed
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+        self.layerdrop = config.layerdrop
+        self.padding_idx = config.pad_token_id
+        self.max_target_positions = config.max_position_embeddings
+        self.share_input_output_embed = config.share_input_output_embed
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+        # Self Attention
+        # Before
+        if self.do_layer_norm_before:
+            hidden_states = self.self_attn_layer_norm(hidden_states)
```
```suggestion
            hidden_states = self.self_attn_layer_norm(hidden_states)
            
```

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+        residual = hidden_states
+
+        # Self Attention
+        # Before
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+        )
+        hidden_states = nn.functional.dropout(hidden_states, p=self.dropout, training=self.training)
+        hidden_states = residual + hidden_states
+        # After
```
```suggestion

```

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+        hidden_states_shape = hidden_states.shape
+        hidden_states = hidden_states.reshape(-1, hidden_states.size(-1))
+        residual = hidden_states
+        # Before
```
```suggestion

```

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+
+        hidden_states = (residual + hidden_states).view(hidden_states_shape)
+        # After
+        if not self.do_layer_norm_before:
```
good job on adding those!

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+        hidden_states = nn.functional.dropout(hidden_states, p=self.dropout, training=self.training)
+
+        hidden_states = (residual + hidden_states).view(hidden_states_shape)
+        # After
```
```suggestion

```

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,140 @@
+        attention_dropout=0.0,
+        activation_dropout=0.0,
+        init_std=0.02,
+        scale_embedding=False,
```
```suggestion
```

Also don't think we need that

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,140 @@
+        layerdrop=0.0,
+        activation_function="relu",
+        d_model=768,
+        word_embed_proj_dim=768,
```
Actually can you do the following here: 

Set it to None and let it default to `d_model` if None (that's the case for 4/5 checkpoints). Only for 350m we need to set it

patrickvonplaten commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,140 @@
+        self.init_std = init_std
+        self.layerdrop = layerdrop
+        self.use_cache = use_cache
+        self.scale_embedding = scale_embedding  # scale factor will be sqrt(d_model) if True
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+        self.padding_idx = config.pad_token_id
+        self.max_target_positions = config.max_position_embeddings
+        self.share_input_output_embed = config.share_input_output_embed
+        self.embed_scale = math.sqrt(config.d_model) if config.scale_embedding else 1.0
```
`config.scale_embeddding` is always False so let's delete this param - it's only confusing

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+        self.embed_tokens = nn.Embedding(config.vocab_size, config.word_embed_proj_dim, self.padding_idx)
+
+        if self.padding_idx is not None:
+            num_embeddings = config.max_position_embeddings + 2
```
Let's add a nice comment here (why we do the +2 - it's a hack) you can copy it from https://github.com/huggingface/transformers/blob/c33f6046c3dab8f41bedf893404e6469dea3bce8/src/transformers/models/bart/modeling_bart.py#L126

ArthurZucker replied:
Okay ;)

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+            inputs_embeds = self.embed_tokens(input_ids) * self.embed_scale
+
+        # embed positions
+        attention_mask = (
```
Actually let's maybe make this an if-else :-)

patrickvonplaten commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+        super().__init__(config)
+
+        self.decoder = OPTDecoder(config)
+        # padding_idx, vocab_size = config.pad_token_id, config.vocab_size
```
```suggestion
```

patrickvonplaten committed:
up

patrickvonplaten committed:
more fixes

patrickvonplaten committed:
model & config work

ArthurZucker committed:
Update src/transformers/models/opt/modeling_opt.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ArthurZucker committed:
Update src/transformers/models/opt/modeling_opt.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ArthurZucker committed:
Update src/transformers/models/opt/modeling_opt.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ArthurZucker committed:
added comment on padding hack (+2)

ArthurZucker commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1017 @@
+        if not ((positions is None) or (self.padding_idx is None)):
+            raise ValueError("If positions is pre-computed then padding_idx should not be set.")
+
+        # we cannot use incremental state here because we must be aware of
```
Okay

ArthurZucker committed:
cleaup

ArthurZucker committed:
review update

ArthurZucker committed:
docstring for missing arg

ArthurZucker committed:
Update docs/source/en/model_doc/opt.mdx

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ArthurZucker committed:
Update docs/source/en/model_doc/opt.mdx

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ArthurZucker committed:
Update docs/source/en/model_doc/opt.mdx

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ArthurZucker committed:
Update src/transformers/models/opt/__init__.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ArthurZucker committed:
update pretrained map

ArthurZucker committed:
Merge branch 'opt-350-m' of https://github.com/younesbelkada/transformers into opt-350-m

ArthurZucker committed:
update path and tests

ArthurZucker committed:
make style

ArthurZucker committed:
styling

ArthurZucker committed:
make consistency

patrickvonplaten committed:
add gpt2 tok new

patrickvonplaten committed:
Merge branch 'opt-350-m' of https://github.com/younesbelkada/transformers into opt-350-m

patrickvonplaten committed:
more tok fixes

patrickvonplaten commented on src/transformers/models/auto/tokenization_auto.py:
```diff
@@ -107,6 +107,7 @@
             ("blenderbot", ("BlenderbotTokenizer", "BlenderbotTokenizerFast")),
             ("tapex", ("TapexTokenizer", None)),
             ("bart", ("BartTokenizer", "BartTokenizerFast")),
+            ("opt", ("GPT2Tokenizer", "GPT2TokenizerFast" if is_tokenizers_available() else None)),
```
```suggestion
            ("opt", ("GPT2Tokenizer", None)),
```

patrickvonplaten replied:
Need special converter for EOS prompting -> leaving out for now

patrickvonplaten committed:
Update src/transformers/models/auto/tokenization_auto.py

sgugger commented on docs/source/en/_toctree.yml:
```diff
@@ -268,6 +268,8 @@
       title: Nyströmformer
     - local: model_doc/openai-gpt
       title: OpenAI GPT
+    - local: model_doc/opt
```
The model will also need to be added to the main README.

ArthurZucker replied:
I am not really I know where I should add it, could you help me out?

patrickvonplaten replied:
on it

sgugger commented on docs/source/en/model_doc/opt.mdx:
```diff
@@ -0,0 +1,54 @@
+
+If you want to use the `opt-350m`, you need to set the `do_layer_norm_before` parameter to `False`. 
+
+This model was contributed by [Arthur Zucker](https://huggingface.co/ArthurZ) and [Younes Belkada](https://huggingface.co/ybelkada), [Patrick Von Platen](https://huggingface.co/patrickvonplaten).
```
```suggestion
This model was contributed by [Arthur Zucker](https://huggingface.co/ArthurZ), [Younes Belkada](https://huggingface.co/ybelkada), and [Patrick Von Platen](https://huggingface.co/patrickvonplaten).
```

sgugger commented on docs/source/en/model_doc/opt.mdx:
```diff
@@ -0,0 +1,54 @@
+## OPTConfig
+
+[[autodoc]] OPTConfig
+    - all
+
```
```suggestion
```
all is not necessary.

sgugger commented on docs/source/en/model_doc/opt.mdx:
```diff
@@ -0,0 +1,54 @@
+[[autodoc]] OPTModel
+    - forward
+
+## OPTPretrainedModel
+
+[[autodoc]] OPTPretrainedModel
+    - all
+
```
```suggestion
```

We don't document this class in general, as there is no useful docstrings.

We should document `OPTModel` however.

patrickvonplaten commented on src/transformers/models/gpt2/tokenization_gpt2.py:
```diff
@@ -162,20 +162,26 @@ def __init__(
         unk_token="<|endoftext|>",
         bos_token="<|endoftext|>",
         eos_token="<|endoftext|>",
+        pad_token=None,
```
Also checked that GPT2 is not affected, but would be great if you could double-check @LysandreJik @sgugger

sgugger commented on src/transformers/models/auto/tokenization_auto.py:
```diff
@@ -107,6 +107,7 @@
             ("blenderbot", ("BlenderbotTokenizer", "BlenderbotTokenizerFast")),
             ("tapex", ("TapexTokenizer", None)),
             ("bart", ("BartTokenizer", "BartTokenizerFast")),
+            ("opt", ("GPT2Tokenizer", None)),
```
LGTM!

sgugger commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,138 @@
+logger = logging.get_logger(__name__)
+
+OPT_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "opt-125m": "https://huggingface.co/facebook/opt-125m/blob/main/config.json",
+    "opt-350m": "https://huggingface.co/facebook/opt-350m/blob/main/config.json",
+    "opt-1.3b": "https://huggingface.co/facebook/opt-1.3b/blob/main/config.json",
+    "opt-2.7b": "https://huggingface.co/facebook/opt-2.7b/blob/main/config.json",
+    "opt-6.7b": "https://huggingface.co/facebook/opt-6.7b/blob/main/config.json",
+    "opt-13b": "https://huggingface.co/facebook/opt-13b/blob/main/config.json",
```
The identifiers are all missing the "facebook" namespace.

ArthurZucker replied:
Fixing!

sgugger commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,138 @@
+    This is the configuration class to store the configuration of a [`OPTModel`]. It is used to instantiate a OPT model
+    according to the specified arguments, defining the model architecture. Instantiating a configuration with the
+    defaults will yield a similar configuration to that of the OPT
+    [facebook/opt-large](https://huggingface.co/facebook/opt-large) architecture.
```
This doesn't match the names above.

sgugger commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1006 @@
+# coding=utf-8
+# Copyright 2021 The Fairseq Authors and The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2022 The Fairseq Authors and The HuggingFace Inc. team. All rights reserved.
```

sgugger commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1006 @@
+        else:
+            self.max_positions = self.num_embeddings
+
+    def forward(
+        self,
+        attention_mask: Tensor,
+        positions: Optional[Tensor] = None,
+    ):
```
```suggestion
    def forward(self, attention_mask: Tensor, positions: Optional[Tensor] = None):
```
Fits in one line.

sgugger commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1006 @@
+        attention_mask: Tensor,
+        positions: Optional[Tensor] = None,
+    ):
+        """attention_masks is expected to be of size [bsz x seqlen]."""
```
```suggestion
        # attention_masks is expected to be of size [batch_size x seq_len].
```
This is not a docstring, more a comment

sgugger commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1006 @@
+            Indices of input sequence tokens in the vocabulary. Padding will be ignored by default should you provide
+            it.
+
+            Indices can be obtained using [`OPTTokenizer`]. See [`PreTrainedTokenizer.encode`] and
+            [`PreTrainedTokenizer.__call__`] for details.
```
```suggestion
            Indices can be obtained using [`GPT2Tokenizer`]. See [`PreTrainedTokenizer.encode`] and
            [`PreTrainedTokenizer.__call__`] for details.
```

sgugger commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1006 @@
+
+            If `past_key_values` are used, the user can optionally input only the last `decoder_input_ids` (those that
+            don't have their past key value states given to this model) of shape `(batch_size, 1)` instead of all
+            `decoder_input_ids` of shape `(batch_size, sequence_length)`. inputs_embeds (`torch.FloatTensor` of shape
+            `(batch_size, sequence_length, hidden_size)`, *optional*): Optionally, instead of passing `input_ids` you
```
Some bad formatting here, `input_embeds` need to go on a new unindented line.
```suggestion
            `decoder_input_ids` of shape `(batch_size, sequence_length)`.
        inputs_embeds (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`, *optional*):
            Optionally, instead of passing `input_ids` you
```

sgugger commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1006 @@
+            num_embeddings = config.max_position_embeddings + 2
+
+        self.embed_positions = OPTLearnedPositionalEmbedding(
+            num_embeddings,
+            config.d_model,
+            self.padding_idx,
```
```suggestion
            num_embeddings, config.d_model, self.padding_idx
```

sgugger commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1006 @@
+        else:
+            self.project_in = None
+
+        # Think we can delete this
+        #        self.layer_norm = nn.LayerNorm(config.d_model) if config.decoder_layernorm else None
```
Let's delete it then ;-)

sgugger commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1006 @@
+
+                If `past_key_values` are used, the user can optionally input only the last `decoder_input_ids` (those
+                that don't have their past key value states given to this model) of shape `(batch_size, 1)` instead of
+                all `decoder_input_ids` of shape `(batch_size, sequence_length)`. inputs_embeds (`torch.FloatTensor` of
+                shape `(batch_size, sequence_length, hidden_size)`, *optional*): Optionally, instead of passing
```
As above, input_embeds is badly formatted

sgugger commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1006 @@
+            if output_attentions:
+                all_self_attns += (layer_outputs[1],)
+
+        #        if self.layer_norm:
+        #            hidden_states = self.layer_norm(hidden_states)
```
To delete?

ArthurZucker replied:
Yes!

sgugger commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1006 @@
+                If `past_key_values` are used, the user can optionally input only the last `decoder_input_ids` (those
+                that don't have their past key value states given to this model) of shape `(batch_size, 1)` instead of
+                all `decoder_input_ids` of shape `(batch_size, sequence_length)`.
+            labels (`torch.LongTensor` of shape `(batch_size, sequence_length)`, *optional*):
```
The documentation of `input_embeds` is missing here.

sgugger commented on tests/models/opt/test_modeling_opt.py:
```diff
@@ -0,0 +1,367 @@
+            "Paris is the capital of France and",
+            "Computers and mobile phones have taken",
+        ]
+        NEXT_TOKENS = [
+            3392,
+            764,
+            5,
+            81,
+        ]
```
```suggestion
        NEXT_TOKENS = [3392, 764, 5, 81]
```

sgugger reviewed (approved):
Thanks a lot for working so fast on this new model integration!

ArthurZucker committed:
Update docs/source/en/model_doc/opt.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
Update docs/source/en/model_doc/opt.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
Update docs/source/en/model_doc/opt.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
Update src/transformers/models/opt/modeling_opt.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
Update tests/models/opt/test_modeling_opt.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
Update src/transformers/models/opt/modeling_opt.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
Update src/transformers/models/opt/modeling_opt.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
Update src/transformers/models/opt/modeling_opt.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
Update src/transformers/models/opt/modeling_opt.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
Update src/transformers/models/opt/modeling_opt.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
Update based on reviews

ArthurZucker committed:
Merge branch 'opt-350-m' of https://github.com/younesbelkada/transformers into opt-350-m

LysandreJik commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,138 @@
+    documentation from [`PretrainedConfig`] for more information.
+
+
+    Args:
```
Documented args are in the wrong ordered when compared to the init method

LysandreJik commented on src/transformers/models/opt/configuration_opt.py:
```diff
@@ -0,0 +1,138 @@
+        vocab_size (`int`, *optional*, defaults to 50272):
+            Vocabulary size of the OPT model. Defines the number of different tokens that can be represented by the
+            `inputs_ids` passed when calling [`OPTModel`]
+        d_model (`int`, *optional*, defaults to 768):
```
This could be `hidden_size` to be similar to most other models

patrickvonplaten replied:
Sounds good to me! Small drawback is that it'll make it harder to compare original codebase, but think unified API def outweighs  here

LysandreJik commented on src/transformers/models/opt/configuration_opt.py:
````diff
@@ -0,0 +1,138 @@
+    ```"""
+    model_type = "opt"
+    keys_to_ignore_at_inference = ["past_key_values"]
+    attribute_map = {"hidden_size": "d_model", "num_hidden_layers": "num_hidden_layers"}
````
Is the `num_hidden_layers` necessary?

LysandreJik commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1001 @@
+        """
+        Args:
+            hidden_states (`torch.FloatTensor`): input to the layer of shape `(batch, seq_len, embed_dim)`
+            attention_mask (`torch.FloatTensor`): attention mask of size
```
```suggestion
            attention_mask (`torch.FloatTensor`, *optional*): attention mask of size
```

LysandreJik commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1001 @@
+            hidden_states (`torch.FloatTensor`): input to the layer of shape `(batch, seq_len, embed_dim)`
+            attention_mask (`torch.FloatTensor`): attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            layer_head_mask (`torch.FloatTensor`): mask for attention heads in a given layer of size
```
```suggestion
            layer_head_mask (`torch.FloatTensor`, *optional*): mask for attention heads in a given layer of size
```

LysandreJik commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1001 @@
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            layer_head_mask (`torch.FloatTensor`): mask for attention heads in a given layer of size
+                `(encoder_attention_heads,)`.
+            past_key_value (`Tuple(torch.FloatTensor)`): cached past key and value projection states
```
```suggestion
            past_key_value (`Tuple(torch.FloatTensor)`, *optional*): cached past key and value projection states
```

LysandreJik commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -0,0 +1,1001 @@
+            layer_head_mask (`torch.FloatTensor`): mask for attention heads in a given layer of size
+                `(encoder_attention_heads,)`.
+            past_key_value (`Tuple(torch.FloatTensor)`): cached past key and value projection states
+            output_attentions (`bool`, *optional*):
+                Whether or not to return the attentions tensors of all attention layers. See `attentions` under
+                returned tensors for more detail.
```
This is at the wrong place, it should be placed before `use_cache` according to the docstring. `use_cache` doesn't seem to be actually documented.

LysandreJik reviewed (approved):
LGTM! Thanks for working on it, @younesbelkada, @ArthurZucker, @patrickvonplaten!

patrickvonplaten committed:
Apply suggestions from code review

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>

patrickvonplaten committed:
make style

patrickvonplaten committed:
make tokenizer auto tests pass

patrickvonplaten committed:
apply Lysandre suggestion

patrickvonplaten committed:
Merge branch 'main' of https://github.com/huggingface/transformers into opt-350-m

patrickvonplaten committed:
finish tests

patrickvonplaten committed:
add some good tokenizer tests

ArthurZucker commented:
Should we merge? 🤗

patrickvonplaten commented:
Actually one sec, I'll update the model docs a bit again

patrickvonplaten committed:
improve docs slighly

patrickvonplaten commented on src/transformers/models/auto/tokenization_auto.py:
```diff
@@ -137,6 +137,7 @@
             ("openai-gpt", ("OpenAIGPTTokenizer", "OpenAIGPTTokenizerFast" if is_tokenizers_available() else None)),
             ("gpt2", ("GPT2Tokenizer", "GPT2TokenizerFast" if is_tokenizers_available() else None)),
             ("gptj", ("GPT2Tokenizer", "GPT2TokenizerFast" if is_tokenizers_available() else None)),
+            ("opt", ("GPT2Tokenizer", None)),
```
Need to add a fast tokenizer in a follow-up PR that is able to prepend the bos_token at the beginning. For this a new converter needs to be written.
